File: medium/208. Implement Trie (Prefix Tree)/208. Implement Trie (Prefix Tree).py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.isWordsEnd = False


class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        curr = self.root

        for letter in word:
            if letter not in curr.children:
                curr.children[letter] = TrieNode()
            curr = curr.children[letter]

        curr.isWordsEnd = True

    def search(self, word: str) -> bool:
        curr = self.root
        i = 0

        while i < len(word) and word[i] in curr.children:
            curr = curr.children[word[i]]
            i += 1

        return curr.isWordsEnd and i == len(word)

    def startsWith(self, prefix: str) -> bool:
        curr = self.root
        i = 0

        while i < len(prefix) and prefix[i] in curr.children:
            curr = curr.children[prefix[i]]
            i += 1

        return i == len(prefix)


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: medium/208. Implement Trie (Prefix Tree)/208. Implement Trie (Prefix Tree).ts
class TrieNode {
    children: { [key in string]: TrieNode }
    isEnd: boolean

    constructor() {
        this.children = {}
        this.isEnd = false
    }
}

class Trie {
    root: TrieNode

    constructor() {
        this.root = new TrieNode()
    }

    insert(word: string): void {
        let current = this.root

        for (const letter of word) {
            if (!(letter in current.children))
                current.children[letter] = new TrieNode()

            current = current.children[letter]
        }

        current.isEnd = true
    }

    search(word: string): boolean {
        let current = this.root
        let i = 0

        while (i < word.length && word[i] in current.children) {
            current = current.children[word[i]]
            i += 1
        }

        return word.length === i && current.isEnd
    }

    startsWith(prefix: string): boolean {
        let current = this.root
        let i = 0

        while (i < prefix.length && prefix[i] in current.children) {
            current = current.children[prefix[i]]
            i += 1
        }

        return prefix.length === i
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * var obj = new Trie()
 * obj.insert(word)
 * var param_2 = obj.search(word)
 * var param_3 = obj.startsWith(prefix)
 */